src/Compilers/CSharp/Portable/SourceGeneration/IndentingStringBuilder.cs

```diff
@@ -0,0 +1,497 @@
+        this.Builder.Append(_endOfLine);
+        return this;
+    }
+
+    /// <summary>
+    /// Appends a single line to the underlying buffer.  Indentation is written out if the underlying buffer
+    /// is at the start of a line.
+    /// </summary>
+    private void AppendSingleLine(ReadOnlySpan<char> line, string? originalLine, bool skipIndent)
+    {
+        if (line.Length == 0)
+            return;
+
+        var builder = this.Builder;
+
+        if (!skipIndent && !IsEndOfLineCharacter(line[0]))
+        {
+            if (builder.Length == 0 || IsEndOfLineCharacter(builder[^1]))
+                builder.Append(_currentIndentation);
+        }
+
+#if NET
+        builder.Append(line);
+#else
+        if (originalLine != null)
+        {
+            builder.Append(originalLine);
+        }
+        else
+        {
+            foreach (var ch in line)
+                builder.Append(ch);
+        }
+#endif
+    }
+
+    private static bool IsEndOfLineCharacter(char ch)
+        => EndOfLineCharacters.IndexOf(ch) >= 0;
+
+    /// <summary>
+    /// Writes content to the underlying buffer.  If the buffer is at the start of a line, then indentation will be
+    /// appended first before the content.  By default, for performance reasons, the content is assumed to contain no
+    /// end of line characters in it.  If the content may contain end of line characters, then <see langword="true"/>
+    /// should be passed in for <paramref name="splitContent"/>.  This will cause the provided content to be split into
+    /// constituent lines, with each line being appended one at a time.
+    /// </summary>
+    /// <param name="skipIndent">If true, will not indent content, even if starting a new line.</param>
+    public IndentingStringBuilder Write(string content, bool splitContent = false, bool skipIndent = false)
+        => Write(content.AsSpan(), content, splitContent, skipIndent);
+
+    /// <inheritdoc cref="Write(string, bool, bool)"/>
+    public IndentingStringBuilder Write(ReadOnlySpan<char> content, bool splitContent = false, bool skipIndent = false)
+        => Write(content, originalString: null, splitContent, skipIndent);
+
+    /// <inheritdoc cref="Write(string, bool, bool)"/>
+    private IndentingStringBuilder Write(ReadOnlySpan<char> content, string? originalString, bool splitContent, bool skipIndent = false)
+    {
+        if (splitContent)
+        {
+            while (content.Length > 0)
+            {
+                var endOfLineIndex = content.IndexOfAny(EndOfLineCharacters);
+                if (endOfLineIndex < 0)
+                {
+                    // no new line, append the rest of the content to the buffer.
+                    AppendSingleLine(content, originalLine: null, skipIndent);
+                }
+                else
+                {
+                    while (endOfLineIndex < content.Length & IsEndOfLineCharacter(content[endOfLineIndex + 1]))
+                        endOfLineIndex++;
```

Comment:
📝 This does not correctly handle inputs with an intentional blank line.

```suggestion
                    if (endOfLineIndex < content.Length && content[endOfLineIndex] == '\r' && content[endOfLineIndex + 1] == '\n')
                        endOfLineIndex++;
```

Reply:
not sure what you mean.  this simply finds the end of the line to append.  even if there are multiple newlines charactesr, they will all be appended, inserting any intended blank lines in the input.

Reply:
put another way, this is finding the first *non-newline* character.  since that corresponds to the text that would need to be indented.

Reply:
I interpreted splitLines to mean that it writes the content as lines using the builder's understanding of the EOL sequence and not the input string.

Reply:
No.  Splitlines just means: "i'm giving you content containing newlines, please break it up for me so i don't have to".  

Generally, we would always do this.  But this allows fast pathing where in teh common case (people are just writing content without newlines in it) we don't do any processing.